sympy/solvers/tests/test_solveset.py

```diff
@@ -1943,3 +1849,110 @@ def test_solve_logarithm_reducible_to_single_instance():
         Eq(log(x)*log(y), 0), x) == log(y**log(x))
 
 # end of logarithmic tests
+
+
+# lambert tests
+
+def test_solve_lambert():
+    a = Symbol('a', real=True)
+    x = Symbol('x')
+    assert solveset_real(3*log(x) - x*log(3), x) == FiniteSet(
+        -3*LambertW(-log(3)/3)/log(3),
+        -3*LambertW(-log(3)/3, -1)/log(3))
+
+    assert solveset_real(3*log(a**(3*x + 5)) + a**(3*x + 5), x) == \
+        FiniteSet((-log(a**5) - LambertW(S(1)/3))/(3*log(a)))
+
+    assert solveset_real(exp(x) + x, x) == FiniteSet(-LambertW(1))
+    assert solveset_real(x + 2**x, x) == \
+        FiniteSet(-LambertW(log(2))/log(2))
+
+    # issue 4739
+    ans = solveset_real(3*x + 5 + 2**(-5*x + 3), x)
+    assert ans == FiniteSet(-Rational(5, 3) +
+                            LambertW(-10240*2**(S(1)/3)*log(2)/3)/(5*log(2)))
+
+    assert solveset_real(5*x - 1 + 3*exp(2 - 7*x), x) == \
+        FiniteSet(Rational(1, 5) + LambertW(-21*exp(Rational(3, 5))/5)/7)
+
+    a = Symbol('a')
+    assert solveset_real(-a*x + 2*x*log(x), x) == \
+        Union(FiniteSet(0), Intersection(S.Reals, FiniteSet(exp(a/2))))
+
+    assert solveset_real(a/x + exp(x/2), x) == \
+        FiniteSet(2*LambertW(-a/2)) - FiniteSet(0)
+
+    assert solveset_real(log(log(x - 3)) + log(x-3), x) == FiniteSet(
+        exp(LambertW(1)) + 3)
+    assert solveset_real(x*exp(x) - 1, x) == FiniteSet(LambertW(1))
+
+    eq = (x*exp(x) - 3).subs(x, x*exp(x))
+    assert solveset_real(eq, x) == \
+        FiniteSet(LambertW(3*exp(-LambertW(3))))
+
+    assert solveset_real(LambertW(2*x) - y, x) == FiniteSet(
+        y*exp(y)/2)
+    # check collection
+    b = Symbol('b')
+    eq = 3*log(a**(3*x + 5)) + b*log(a**(3*x + 5)) + a**(3*x + 5)
+    assert solveset_real(eq, x) == FiniteSet(
+        (-log(a**5) - LambertW(S(1)/(b + 3)))/(3*log(a)))
+
+    p = symbols('p', positive=True)
+    eq = 3*log(p**(3*x + 5)) + p**(3*x + 5)
+    assert solveset(eq, x) == FiniteSet(log(3**(S(1)/3)*LambertW(S(1)/3)**(S(1)/3)/p**(S(5)/3))/log(p),
```

Comment:
I am not sure about the correctness of this test in the complex domain. Also, I don't think principal value should be considered while solving in `solveset`